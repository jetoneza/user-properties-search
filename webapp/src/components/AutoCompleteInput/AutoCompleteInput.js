import React from 'react';
import PropTypes from 'prop-types';
import { Search, Grid } from 'semantic-ui-react';

import { imageUrlFaker } from '../../utils/mocks';
import searchResultBuilder from '../../utils/searchResultBuilder';
import { USER_TYPE, PROPERTY_TYPE } from '../../utils/constants';
import debounce from '../../utils/debounce';

export default class AutoCompleteInput extends React.Component {
  static propTypes = {
    onResultSelect: PropTypes.func.isRequired,
    onSearchChange: PropTypes.func.isRequired,
    results: PropTypes.shape({
      users: PropTypes.array.isRequired,
      properties: PropTypes.array.isRequired,
    }).isRequired,
  }

  constructor(props) {
    super(props);

    this.state = {
      value: '',
    };

    this.debouncedChangeHandler = debounce(this.autoCompleteSearch, 500);
  }

  handleResultSelect = (e, { result }) => {
    this.setState({ value: result.title });

    this.props.onResultSelect(result);
  }

  autoCompleteSearch = (e) => {
    this.props.onSearchChange(e.target.value);
  }

  handleChange = (e) => {
    e.persist();

    this.setState({ value: e.target.value });

    this.debouncedChangeHandler(e);
  }

  buildResults = (results) => {
    const { users, properties } = results;

    const categories = [
      {
        key: USER_TYPE.KEY,
        name: USER_TYPE.NAME,
        results: users,
        resultBuilder: (user) => ({
          type: USER_TYPE.SINGLE_KEY,
          title: `${user.firstName} ${user.lastName}`,
          description: `${user.properties.length} Listing/s`,
          image: imageUrlFaker('face', user.id),
          properties: user.properties,
        }),
      },
      {
        key: PROPERTY_TYPE.KEY,
        name: PROPERTY_TYPE.NAME,
        results: properties,
        resultBuilder: (property) => ({
          type: PROPERTY_TYPE.SINGLE_KEY,
          title: property.street,
          description: `${property.city}, ${property.state}`,
          price: `$${property.rent}`,
          image: imageUrlFaker('house', property.id),
          user: property.user,
        }),
      },
    ];

    return searchResultBuilder(categories);
  }

  render() {
    const { loading, results } = this.props;
    const { value } = this.state;

    return (
      <Grid centered>
        <Grid.Column width={16}>
          <Search
            category
            loading={loading}
            onResultSelect={this.handleResultSelect}
            onSearchChange={this.handleChange}
            results={this.buildResults(results)}
            value={value}
          />
        </Grid.Column>
      </Grid>
    );
  }
}
