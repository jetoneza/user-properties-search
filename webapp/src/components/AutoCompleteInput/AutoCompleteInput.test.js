import React from 'react';
import renderer from 'react-test-renderer';
import sinon from 'sinon';

import { mount } from 'enzyme';

import AutoCompleteInput from './AutoCompleteInput';

describe('<AutoCompleteInput /> test', () => {
  const onResultSelect = jest.fn();
  const onSearchChange = jest.fn();

  const props = {
    onResultSelect,
    onSearchChange,
    results: {
      users: [],
      properties: [],
    }
  };

  test('Should update state on input change', () => {
    const wrapper = mount(<AutoCompleteInput {...props} />);

    const input = wrapper.find('input').first();

    input.simulate('change', { target: { value: 'John Doe' } });

    expect(wrapper.state().value).toEqual('John Doe');
  });

  test('Should render input', () => {
    const component = <AutoCompleteInput {...props} />;
    const wrapper = mount(component);

    const renderedComponent = renderer.create(component);
    const tree = renderedComponent.toJSON();

    expect(tree).toMatchSnapshot();
  });
});


