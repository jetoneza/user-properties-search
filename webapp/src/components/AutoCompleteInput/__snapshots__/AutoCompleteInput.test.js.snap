// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AutoCompleteInput /> test Should render input 1`] = `
<div
  className="ui centered grid"
>
  <div
    className="sixteen wide column"
  >
    <div
      className="ui category search"
      onBlur={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
    >
      <div
        className="ui icon input"
      >
        <input
          autoComplete="off"
          className="prompt"
          disabled={undefined}
          onChange={[Function]}
          onClick={[Function]}
          tabIndex="0"
          type="text"
          value=""
        />
        <i
          aria-hidden="true"
          className="search icon"
        />
      </div>
      <div
        className="results transition"
      >
        <div
          className="message empty"
        >
          <div
            className="header"
          >
            No results found.
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
